---
title: "Initial analysis"
date: "2023-04-12"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
---

# Background

We have cohort of samples profiled for proteomics, genomics and metylation. We want to create a harmonized clinical file for all these patients.

In regards to proteomics. Samples were profiled ;

1) Psomagen cohort
2) Olink 2023 cohort (Q-13356) - Also called Project 2
3) Olink 2024 cohort (Q-15806) - Also called Project 4

In addition, some patients were also profiled by genomics (KLK2 SNPs) and methylation.

# Objectives

1. Gather data
2. Clean up clinical data
3. Viz of data

# Conclusion 

We extracted the metadata from 3 proteomics experiment. Identified samples to be used in future analysis. We have the final patient and sample numbers. 

# Pre-processing 

## Loading packages

```{r, message=FALSE, warning=FALSE}
library(here)
library(tidyverse)
library(readxl)   
library(readr)
library(stringr)

library(tidyverse)
library(ggplot2)
library(VennDiagram)
library(ggvenn)
```


## Directories

```{r}
wd <- list()
wd$main <- here()
wd$data <- file.path(wd$main, "data")
wd$manishData <- file.path(wd$data, "raw/data_fromManish")
wd$d2024 <- file.path(wd$data, "updated_2024")
wd$d2024_npx <- file.path(wd$d2024, "NPX_data")
wd$output <- file.path(wd$main, "output")
wd$script <- file.path(wd$main, "scripts")

wd$outData <- file.path(wd$output, "data")
wd$outCurr <- file.path(wd$output, "01_InitialAnalysis")
```


Create directories

```{r}
if (!file.exists(wd$outCurr)) {
  dir.create(wd$outCurr)
} else {
  print("Directory already exists")
}
```

Load functions

```{r}
source(file.path(wd$script, "functions.R"))
```


# Objective 1

**Gather data**

We start with the information that there are 312 samples sent for Proteomics profiling. For each patient, we have the HCI collection ID (HCI_cID) and the sample ID by Olink. 

We read the master file of all the HCI_cID into R

```{r}
df_hci_id <- read.csv(file.path(wd$d2024, "clinical_data/MetaData_Proteomics.csv"))
nrow(df_hci_id)
```

We also specify in the clinical file provided by Enos and Becky. This file is available in the cloud as well at - 

```{r}
meta_file <- file.path(wd$d2024, "clinical_data/Proteomic Demographic 06_23_25.xlsx")
```

## Cohort summary

Enos provided us with an updated clinical data. We stored this as `meta_file`.

Tab 1 contains cohort A. 
Tab 2 contains cohort B. 
Tab 3 contains cohort C. 

### Cohort A

For cohort A, lets get the HCI_cID and collection date

```{r}
df_clin <- read_xlsx(meta_file, sheet = 1)

dd <- df_clin

df_a <- data.frame(
    mrn = dd$`MRN (UUHSC)`,
    age = dd$Age_at_time_of_enrollment,
    HCI_cID = c(dd$Sample1_HCI_Num, dd$Sample2_HCI_Num),
    date = c(dd$Sample1_date, dd$`Sample 2 Start date`),
    psa = c(dd$sample1_psa, dd$sample2_psa)
  ) %>% filter(!is.na(HCI_cID)) %>% 
  mutate(cohort = "A") %>% 
  filter(HCI_cID %in% df_hci_id$HCI_cID)

df_a2 <- data.frame(
    mrn = dd$`MRN (UUHSC)`,
    age = dd$Age_at_time_of_enrollment,
    HCI_cID = c(dd$Sample1_HCI_Num, dd$Sample2_HCI_Num),
    date = c(dd$Sample1_date, dd$`Sample 2 Start date`),
    psa = c(dd$sample1_psa, dd$sample2_psa),
    alk_phos = c(dd$sample1_alk_phos, dd$sample2_alk_phos),
    testo = NA,  # No testosterone data for Cohort A
    hemoGlo = c(dd$`Hemoglobin at s1`, rep(NA, length(dd$Sample1_HCI_Num))),  # Only sample 1 has hemoglobin data
    surgery_type = dd$Surgery_Type,
    glsn_diag = dd$diag_glsn,
    death_status = dd$`Deceased?`)  %>%
  filter(!is.na(HCI_cID)) %>%
  mutate(cohort = "A") %>%
  filter(HCI_cID %in% df_hci_id$HCI_cID)
```

### Cohort B

For cohort B, lets get the HCI_cID and collection date. We also want to add the treatment date. 

```{r}
df_clin <- read_xlsx(meta_file, sheet = 2)

dd <- df_clin
HCI_cID = c(dd$Sample1_HCI_Num, dd$Sample2_HCI_Num, dd$Sample3_HCI_Num)


df_b <- data.frame(
  mrn = dd$`MRN (UUHSC)`,
  age = dd$Age_at_time_of_enrollment,
  HCI_cID = c(dd$Sample1_HCI_Num, dd$Sample2_HCI_Num, dd$Sample3_HCI_Num),
  date = c(dd$Sample1_date, dd$`Sample 2 Start date`, dd$`sample3 date`),
  psa = c(dd$sample1_psa, dd$sample2_psa, dd$sample3_psa),
  testo = c(dd$sample1_testos, dd$`Sample 2 Testerone`, dd$`Sample 3 Testosterone`),
  hemoGlo = c(dd$`Hemoglobin at s1`, rep(NA, (length(HCI_cID) - length(dd$`Hemoglobin at s1`)  ) )),
  surgery_type = dd$Surgery_Type,
  glsn_diag = dd$diag_glsn,
  mets_stage = dd$`De Novo Mets`,
  mets_volume = dd$chaarted_volume_at_mhspc,
  txt_ori = dd$txt_stat,
  death_date = c(dd$`Death date`),
  death_status = dd$`Deceased?`,
  # Info on HSPC treatment
  txt_nhspc_1st_type = dd$`1st_nmhspc_rx`,
  txt_nhspc_1st_date = dd$`1st_nmhspc_rx_date`,
  txt_mhspc_1st_type = dd$`1st_mhspc_rx`,
  txt_mhspc_1st_date = dd$`1st_mhspc_rx_date`,
  txt_mhspc_2nd_type = dd$`2nd_mhspc_rx`,
  txt_mhspc_2nd_date = dd$`2nd_mhspc_rx_date`,
  # Info on CRPC progression
  progressed_to_crpc = dd$`Progressed to CRPC`,
  crpc_biochem_date = dd$biochem_crpc_date,
  txt_nmcrpc_1st_date = dd$`1st_nmcrpc_rx_date`,
  txt_mcrpc_1st_date = dd$`1st_mcrpc_rx_date`,
  txt_mcrpc_2nd_date = dd$`2nd_mcrpc_rx_date`
) %>%
  filter(!is.na(HCI_cID)) %>%
  mutate(
    cohort = "B",
    psa = as.numeric(gsub("<", "", psa)),
    testo = as.numeric(gsub("<", "", testo))
  ) %>%
  filter(HCI_cID %in% df_hci_id$HCI_cID)
```

#### ADT treatment

We want to figure out when is the first ADT date that is given. Lets first list all the treatments given and figure out which one is ADT.

```{r}
# Create columns indicating if ADT was given and the earliest date of ADT given for each patient
df_b2 <- df_b %>%
  mutate(
    # Identify if each treatment type is ADT
    nhspc_adt = str_detect(tolower(txt_nhspc_1st_type), "adt"),
    mhspc1_adt = str_detect(tolower(txt_mhspc_1st_type), "adt"),
    mhspc2_adt = str_detect(tolower(txt_mhspc_2nd_type), "adt"),
    # Convert all treatment dates to Date type
    nhspc_adt_date = if_else(nhspc_adt, as.Date(txt_nhspc_1st_date), as.Date(NA)),
    mhspc1_adt_date = if_else(mhspc1_adt, as.Date(txt_mhspc_1st_date), as.Date(NA)),
    mhspc2_adt_date = if_else(mhspc2_adt, as.Date(txt_mhspc_2nd_date), as.Date(NA))
  ) %>%
  rowwise() %>%
  mutate(
    # Was ADT given in any of the three?
    adt_given = any(c(nhspc_adt, mhspc1_adt, mhspc2_adt), na.rm = TRUE),
    # Find the earliest date of ADT given
    adt_earliest_date = min(c(nhspc_adt_date, mhspc1_adt_date, mhspc2_adt_date), na.rm = TRUE)
  ) %>%
  ungroup() %>%
  mutate(
    # If no ADT was given, set adt_earliest_date to NA
    adt_earliest_date = if_else(adt_given, adt_earliest_date, as.Date(NA))
  )

df_b2 <- df_b %>%
  mutate(
    # Identify if each treatment type is ADT
    nhspc_adt = str_detect(tolower(txt_nhspc_1st_type), "adt"),
    mhspc1_adt = str_detect(tolower(txt_mhspc_1st_type), "adt"),
    mhspc2_adt = str_detect(tolower(txt_mhspc_2nd_type), "adt"),
    # Convert all treatment dates to Date type
    nhspc_adt_date = if_else(nhspc_adt, as.Date(txt_nhspc_1st_date), as.Date(NA)),
    mhspc1_adt_date = if_else(mhspc1_adt, as.Date(txt_mhspc_1st_date), as.Date(NA)),
    mhspc2_adt_date = if_else(mhspc2_adt, as.Date(txt_mhspc_2nd_date), as.Date(NA))
  ) %>%
  rowwise() %>%
  mutate(
    # Was ADT given in any of the three?
    adt_given = any(c(nhspc_adt, mhspc1_adt, mhspc2_adt), na.rm = TRUE),
    # Find the earliest and latest date of ADT given
    adt_earliest_date = min(c(nhspc_adt_date, mhspc1_adt_date, mhspc2_adt_date), na.rm = TRUE),
    adt_latest_date   = max(c(nhspc_adt_date, mhspc1_adt_date, mhspc2_adt_date), na.rm = TRUE)
  ) %>%
  ungroup() %>%
  mutate(
    # If no ADT was given, set adt_earliest_date and adt_latest_date to NA
    adt_earliest_date = if_else(adt_given, adt_earliest_date, as.Date(NA)),
    adt_latest_date   = if_else(adt_given, adt_latest_date, as.Date(NA)))
```

Now we want to create a column to indicate if the patient was given ADT within 30 days of the sample collecton date. For samples before the 30 days, label as Pre and after 30 days label as post. 

For some patients (with >90 days since ADT) Bekcy provided us with an updated ADT date as well. We read this updated dates

```{r}
df_adt <- read_xlsx(file.path(wd$d2024, "clinical_data/ADT_timeline_HSPC.xlsx"), sheet=1)
```

Based on the HCI_cID, we want to replace the ADT collection date to Becky's date. We replace the adt_earliest_date with the dates provided by Becky.

```{r}
df_b2 <- df_b2 %>%
  left_join(df_adt %>% select(HCI_cID, Becky_ADT_date), by = "HCI_cID") %>%
  mutate(
    adt_earliest_date = if_else(!is.na(Becky_ADT_date), as.Date(Becky_ADT_date), adt_earliest_date)
  ) %>%
  select(-Becky_ADT_date)
```

Lets calcualte the time of sample collection to ADT

```{r}
df_b2 <- df_b2 %>%
  mutate(
    # Calculate days from sample collection to ADT
    days_to_adt_early = as.numeric(difftime(date, adt_earliest_date, units = "days")),
    days_to_adt_late  = as.numeric(difftime(date, adt_latest_date, units = "days")),
    # Label as Pre if sample is within 30 days before/after earliest ADT, Post if >30 days after earliest ADT
    adt_status_early = case_when(
      is.na(adt_earliest_date) ~ "Pre",                # No ADT date available
      days_to_adt_early >= -30 & days_to_adt_early <= 30 ~ "Pre",  # Within 30 days before/after earliest ADT
      days_to_adt_early > 30 ~ "Post",                       # More than 30 days after earliest ADT
      days_to_adt_early < -30 ~ "Pre",                       # More than 30 days before earliest ADT 
      TRUE ~ NA_character_),
    adt_status_late = case_when(
      is.na(adt_latest_date) ~ "Pre",                # No ADT date available
      days_to_adt_late >= -30 & days_to_adt_late <= 30 ~ "Pre",  # Within 30 days before/after earliest ADT
      days_to_adt_late > 30 ~ "Post",                       # More than 30 days after earliest ADT
      days_to_adt_late < -30 ~ "Pre",                       # More than 30 days before earliest ADT 
      TRUE ~ NA_character_)
  )
```

When we compare if the Pre- or Post- is the same using early or late. 3 patients were different ;

- 2021-1600 
- 2020-3493 
- 2021-2049 

For now, we are using the days_to_adt_early to define the ADT status. Additionally, we are now considering the testosterone level as well. 

Anything labeled initially as "Pre" and have a testo value <50 will be considered "Post" 

Anything labeled initially as "Post" and have a testo value >50 will be considered as "Pre"

```{r}
df_b3 <- df_b2 %>% 
  mutate(
    updated_adt_early = case_when(
      adt_status_early == "Pre" & testo  <= 50 ~ "Post",
      adt_status_early == "Post" & testo  > 50 ~ "Pre",
      TRUE ~ adt_status_early
    )
  )
```

Number of samples in pre- or post-

```{r}
table(df_b3$updated_adt_early)
```

Lets check how many pre- post- pairs we have

```{r}
mrn_pre_post <- df_b3 %>%
  group_by(mrn) %>%
  summarise(
    has_pre = as.integer(any(updated_adt_early == "Pre", na.rm = TRUE)),
    has_post = as.integer(any(updated_adt_early == "Post", na.rm = TRUE))
  )

# View only those with both Pre and Post
mrn_pre_post %>%
  filter(has_pre == 1 & has_post == 1) %>% 
  dim()
```

#### CRPC progression

We want to also identify the earliest date of CRPC progression

```{r}
df_b3 <- df_b3 %>%
  mutate(
    # Convert all relevant CRPC treatment date columns to Date type
    crpc_biochem_date = as.Date(crpc_biochem_date),
    txt_nmcrpc_1st_date = as.Date(txt_nmcrpc_1st_date),
    txt_mcrpc_1st_date = as.Date(txt_mcrpc_1st_date),
    txt_mcrpc_2nd_date = as.Date(txt_mcrpc_2nd_date),
    # Find the earliest date of CRPC progression
    crpc_earliest_date = pmin(
      crpc_biochem_date,
      txt_nmcrpc_1st_date,
      txt_mcrpc_1st_date,
      txt_mcrpc_2nd_date,
      na.rm = TRUE
    )
  )
```


### Cohort C

For cohort C, lets get the HCI_cID and collection date. We also want to add the treatment date. 

```{r}
df_clin <- read_xlsx(meta_file, sheet = 5)

dd <- df_clin
HCI_cID = c(dd$Sample1_HCI_Num, dd$Sample2_HCI_Num, dd$Sample3_HCI_Num)

df_c <- data.frame(
  mrn = dd$`MRN (UUHSC)`,
  age = dd$Age_at_time_of_enrollment,
  HCI_cID = c(dd$Sample1_HCI_Num, dd$Sample2_HCI_Num, dd$Sample3_HCI_Num),
  date = c(dd$Sample1_date, dd$`Sample 2 Start date`, dd$`Sample3_Start date`),
  psa = c(dd$sample1_psa, dd$sample2_psa, dd$sample3_psa),
  ldh = c(dd$sample1_ldh, dd$sample2_ldh, dd$sample3_ldh),
  alk_phos = c(dd$sample1_alk_phos, dd$sample2_alk_phos, dd$sample3_alk_phos),
  testo = c(dd$sample1_testos, dd$`Sample 2 Testos`, rep(NA, (length(HCI_cID) - length(c(dd$`sample1_testos`, dd$`Sample 2 Testos`)  ) ) )),
  hemoGlo = c(dd$`Hemoglobin at s1`, rep(NA, (length(HCI_cID) - length(dd$`Hemoglobin at s1`)  ) )),
  surgery_type = dd$Surgery_Type,
  glsn_diag = dd$diag_glsn,
  mets_stage = dd$`De-Novo Metastatic (Y/N)`,
  death_date = c(dd$`Death date`),
  death_status = dd$`Deceased?`,
  txt_nmcrpc_1st_type = dd$`1st_nmcrpc_rx`,
  txt_nmcrpc_1st_date = dd$`1st_nmcrpc_rx_date`,
  txt_mcrpc_1st_type = dd$`1st_mcrpc_rx`,
  txt_mcrpc_1st_date = dd$`1st_mcrpc_rx_date`,
  txt_mcrpc_2nd_type = dd$`2nd_mcrpc_rx`,
  txt_mcrpc_2nd_date = dd$`2nd_mcrpc_rx_date`
) %>%
  filter(!is.na(HCI_cID)) %>%
  mutate(
    cohort = "C",
    psa = as.numeric(gsub("<", "", psa)),
    ldh = as.numeric(gsub("<", "", ldh)),
    alk_phos = as.numeric(gsub("<", "", alk_phos)),
    testo = as.numeric(gsub("<", "", testo))
  ) %>%
  filter(HCI_cID %in% df_hci_id$HCI_cID)
```

We want to figure out when is the first CRPC treatment is given. Regardless of what CRPC treatment was given

```{r}
# If any treatment is given, find the earliest and latest treatment date.

# Cohort C: Find earliest and latest treatment date (any treatment)
df_c2 <- df_c %>%
  mutate(
    # Convert all treatment dates to Date type
    nmcrpc_date = as.Date(txt_nmcrpc_1st_date),
    mcrpc1_date = as.Date(txt_mcrpc_1st_date),
    mcrpc2_date = as.Date(txt_mcrpc_2nd_date)
  ) %>%
  rowwise() %>%
  mutate(
    # Find the earliest and latest date of any treatment given
    any_treatment_earliest_date = min(c(nmcrpc_date, mcrpc1_date, mcrpc2_date), na.rm = TRUE),
    any_treatment_latest_date   = max(c(nmcrpc_date, mcrpc1_date, mcrpc2_date), na.rm = TRUE),
    # Was any treatment given? (at least one non-NA date)
    any_treatment_given = any(!is.na(c(nmcrpc_date, mcrpc1_date, mcrpc2_date)))
  ) %>%
  ungroup() %>%
  mutate(
    # If no treatment was given, set dates to NA
    any_treatment_earliest_date = if_else(any_treatment_given, any_treatment_earliest_date, as.Date(NA)),
    any_treatment_latest_date   = if_else(any_treatment_given, any_treatment_latest_date, as.Date(NA))
  )
```

Now we want to create a column to indicate if the patient was given mCRPC treament within 30 days of the sample collection date. For samples wihtin 30 days (before or after the 30 days), label as Pre and after 30 day window label as post. 

Lets calcualte the time of sample collection to treatment

```{r}
df_c2 <- df_c2 %>%
  mutate(
    # Calculate days from sample collection to earliest and latest mCRPC (any) treatment
    days_to_mcrpc_early = as.numeric(difftime(date, any_treatment_earliest_date, units = "days")),
    days_to_mcrpc_late  = as.numeric(difftime(date, any_treatment_latest_date, units = "days")),
    # Label as Pre if sample is within 30 days before/after earliest mCRPC treatment, Post if >30 days after earliest mCRPC treatment
    mcrpc_status_early = case_when(
      is.na(any_treatment_earliest_date) ~ "Pre",                # No mCRPC treatment date available
      days_to_mcrpc_early >= -30 & days_to_mcrpc_early <= 30 ~ "Pre",  # Within 30 days before/after earliest mCRPC treatment
      days_to_mcrpc_early > 30 ~ "Post",                       # More than 30 days after earliest mCRPC treatment
      days_to_mcrpc_early < -30 ~ "Pre",                       # More than 30 days before earliest mCRPC treatment 
      TRUE ~ NA_character_),
    mcrpc_status_late = case_when(
      is.na(any_treatment_latest_date) ~ "Pre",                # No mCRPC treatment date available
      days_to_mcrpc_late >= -30 & days_to_mcrpc_late <= 30 ~ "Pre",  # Within 30 days before/after latest mCRPC treatment
      days_to_mcrpc_late > 30 ~ "Post",                       # More than 30 days after latest mCRPC treatment
      days_to_mcrpc_late < -30 ~ "Pre",                       # More than 30 days before latest mCRPC treatment 
      TRUE ~ NA_character_)
  )
```

Number of samples in pre- or post-

```{r}
table(df_c2$mcrpc_status_early)
```

Lets check how many pre- post- pairs we have

```{r}
mrn_pre_post <- df_c2 %>%
  group_by(mrn) %>%
  summarise(
    has_pre = as.integer(any(mcrpc_status_early == "Pre", na.rm = TRUE)),
    has_post = as.integer(any(mcrpc_status_early == "Post", na.rm = TRUE))
  )

# View only those with both Pre and Post
mrn_pre_post %>%
  filter(has_pre == 1 & has_post == 1) %>% 
  dim()
```

